### ISSUE TITLE
Write failures caused by overload not reported 

### ISSUE THREAD

**Issue by salex89** [kairosdb/kairosdb#145]
Hello,

I'm playing quite a bit with KairosDB for a new project and I like it, but I'm encountering an important issue, with fail-safe reporting. I'm writing quite a lot of data, 900k data points per hour on a 3-node Cassandra cluster (2CPU x 2 cores, 16GB RAM). The performance is ok, while KairosDB is working. 

First of all I had to lower the buffer to 300000 and set the write delay to 3000 sec. This is all still fine performance wise and I have stable writes. Issues arise when the cluster is overloaded. The same happens with normal CQL writes, so I know these are the limits of my cluster

The issues I had before tweaking the database are the exceptions only shown in the log, but they are never reported back to the client. I first thought it is a driver issue, but then I looked at the code and realized it is a database issue. The exception in the log looks like this:

    09:37:09.765 [pool-2-thread-1] ERROR [HThriftClient.java:132] - Could not flush transport (to be        expected if the pool is shutting down) in close for client: CassandraClient<192.168.1.100:9160-3>
    org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe
     	at org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:161) ~[libthrift-0.9.1.jar:0.9.1]
	  at org.apache.thrift.transport.TFramedTransport.flush(TFramedTransport.java:158) ~[libthrift-0.9.1.jar:0.9.1]
	at me.prettyprint.cassandra.connection.client.HThriftClient.close(HThriftClient.java:130) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.connection.client.HThriftClient.close(HThriftClient.java:40) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.connection.HConnectionManager.closeClient(HConnectionManager.java:331) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.connection.HConnectionManager.operateWithFailover(HConnectionManager.java:257) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.ExecutingKeyspace.doExecuteOperation(ExecutingKeyspace.java:102) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.MutatorImpl.execute(MutatorImpl.java:265) [hector-core-2.0-0.jar:na]
	at org.kairosdb.datastore.cassandra.WriteBuffer$WriteDataJob.run(WriteBuffer.java:372) [kairosdb-0.9.5-0.1beta.jar:0.9.5-0.1beta.20150126131205]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_25]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_25]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_25]
    Caused by: java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:1.8.0_25]
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109) ~[na:1.8.0_25]
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153) ~[na:1.8.0_25]
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82) ~[na:1.8.0_25]
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140) ~[na:1.8.0_25]
	at org.apache.thrift.transport.TIOStreamTransport.flush(TIOStreamTransport.java:159) ~[libthrift-0.9.1.jar:0.9.1]
	... 13 common frames omitted
    09:37:09.767 [pool-2-thread-1] WARN  [TIOStreamTransport.java:112] - Error closing output stream.
    java.net.SocketException: Socket closed
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:116) ~[na:1.8.0_25]
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153) ~[na:1.8.0_25]
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82) ~[na:1.8.0_25]
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140) ~[na:1.8.0_25]
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158) ~[na:1.8.0_25]
	at org.apache.thrift.transport.TIOStreamTransport.close(TIOStreamTransport.java:110) ~[libthrift-0.9.1.jar:0.9.1]
	at org.apache.thrift.transport.TSocket.close(TSocket.java:194) [libthrift-0.9.1.jar:0.9.1]
	at org.apache.thrift.transport.TFramedTransport.close(TFramedTransport.java:89) [libthrift-0.9.1.jar:0.9.1]
	at me.prettyprint.cassandra.connection.client.HThriftClient.close(HThriftClient.java:135) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.connection.client.HThriftClient.close(HThriftClient.java:40) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.connection.HConnectionManager.closeClient(HConnectionManager.java:331) [hector-core-2.0-0.jar:na]
	at     me.prettyprint.cassandra.connection.HConnectionManager.operateWithFailover(HConnectionManager.java:257) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.ExecutingKeyspace.doExecuteOperation(ExecutingKeyspace.java:102) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.MutatorImpl.execute(MutatorImpl.java:265) [hector-core-2.0-0.jar:na]
	at org.kairosdb.datastore.cassandra.WriteBuffer$WriteDataJob.run(WriteBuffer.java:372) [kairosdb-0.9.5-0.1beta.jar:0.9.5-0.1beta.20150126131205]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_25]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_25]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_25]
    09:37:09.768 [pool-2-thread-1] ERROR [HConnectionManager.java:429] - MARK HOST AS DOWN TRIGGERED for host 192.168.1.100(192.168.1.100):9160
    09:37:09.768 [pool-2-thread-1] ERROR [HConnectionManager.java:433] - Pool state on shutdown: <ConcurrentCassandraClientPoolByHost>:{192.168.1.100(192.168.1.100):9160}; IsActive?: true; Active: 1; Blocked: 0; Idle: 15; NumBeforeExhausted: 49
    09:37:09.769 [pool-2-thread-1] INFO  [ConcurrentHClientPool.java:195] - Shutdown triggered on <ConcurrentCassandraClientPoolByHost>:{192.168.1.100(192.168.1.100):9160}
    09:37:09.770 [pool-2-thread-1] INFO  [ConcurrentHClientPool.java:203] - Shutdown complete on <ConcurrentCassandraClientPoolByHost>:{192.168.1.100(192.168.1.100):9160}
    09:37:09.770 [pool-2-thread-1] INFO  [CassandraHostRetryService.java:70] - Host detected as down was added to retry queue: 192.168.1.100(192.168.1.100):9160
    09:37:09.772 [pool-2-thread-1] WARN  [HConnectionManager.java:302] - Could not fullfill request on this host CassandraClient<192.168.1.100:9160-3>
    09:37:09.772 [pool-2-thread-1] WARN  [HConnectionManager.java:303] - Exception: 
    me.prettyprint.hector.api.exceptions.HectorTransportException: [192.168.1.100(192.168.1.100):9160] org.apache.thrift.transport.TTransportException: java.net.SocketException: Connection reset
	at me.prettyprint.cassandra.service.ExceptionsTranslatorImpl.translate(ExceptionsTranslatorImpl.java:43) ~[hector-core-2.0-0.jar:na]
	at          me.prettyprint.cassandra.connection.HConnectionManager.operateWithFailover(HConnectionManager.java:249) ~[hector-core-2.0-0.jar:na]
	at     me.prettyprint.cassandra.model.ExecutingKeyspace.doExecuteOperation(ExecutingKeyspace.java:102) [hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.MutatorImpl.execute(MutatorImpl.java:265) [hector-core-2.0-0.jar:na]
	at org.kairosdb.datastore.cassandra.WriteBuffer$WriteDataJob.run(WriteBuffer.java:372) [kairosdb-0.9.5-0.1beta.jar:0.9.5-0.1beta.20150126131205]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_25]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_25]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_25]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_25]
    Caused by: org.apache.thrift.transport.TTransportException: java.net.SocketException: Connection reset
	at org.apache.thrift.transport.TIOStreamTransport.write(TIOStreamTransport.java:147) ~[libthrift-0.9.1.jar:0.9.1]
	at org.apache.thrift.transport.TFramedTransport.flush(TFramedTransport.java:157) ~[libthrift-0.9.1.jar:0.9.1]
	at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:65) ~[libthrift-0.9.1.jar:0.9.1]
	at org.apache.cassandra.thrift.Cassandra$Client.send_batch_mutate(Cassandra.java:1063) ~[cassandra-thrift-2.0.4.jar:2.0.4]
	at org.apache.cassandra.thrift.Cassandra$Client.batch_mutate(Cassandra.java:1054) ~[cassandra-thrift-2.0.4.jar:2.0.4]
	at me.prettyprint.cassandra.model.MutatorImpl$3.execute(MutatorImpl.java:268) ~[hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.model.MutatorImpl$3.execute(MutatorImpl.java:265) ~[hector-core-2.0-0.jar:na]
	at me.prettyprint.cassandra.service.Operation.executeAndSetResult(Operation.java:104) ~[hector-core-2.0-0.jar:na]
	at        me.prettyprint.cassandra.connection.HConnectionManager.operateWithFailover(HConnectionManager.java:241) ~[hector-core-2.0-0.jar:na]
	... 8 common frames omitted
    Caused by: java.net.SocketException: Connection reset
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:113) ~[na:1.8.0_25]
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153) ~[na:1.8.0_25]
	at java.io.BufferedOutputStream.write(BufferedOutputStream.java:122) ~[na:1.8.0_25]
	at org.apache.thrift.transport.TIOStreamTransport.write(TIOStreamTransport.java:145) ~[libthrift-0.9.1.jar:0.9.1]
	... 16 common frames omitted
    09:37:09.778 [pool-2-thread-1] INFO  [HConnectionManager.java:415] - Client         CassandraClient<192.168.1.100:9160-3> released to inactive or dead pool. Closing.

[The driver](https://github.com/kairosdb/kairosdb-client), based on HTTP, reports nothing. It just returns 204. I've also used the decompiled version and debugged it and realized that it's the database not reporting anything. It looks like it tries halving the buffer, which is ok, but data is already lost, it does not get written. 

Bottom line, is there any way to fix the database so it reports failed writes, and retry on the client side? It would be even better if it could handle it itself, but that really isn't a requirement. I'm also interested into changing that and sending a pull request, if that will speed things up and I can continue development. 

It's not excluded other failed writes for other reasons also don't report anything. 

Best regards

**Comment by brianhks** [comment]
Thanks for the feedback. The real issue here is that writing data to Cassandra happens on a different thread from the incoming request.  Data is buffered up in chunks and then batched to Cassandra.  So if we cannot send data to Cassandra it is long after the http message has left the building.

Not sure how to report that situation.  I'm open to suggestions.

It is also worthy to note that because it gets an exception the data is not lost.  The code waits and then tries again until it is sent.

Reducing the buffer size is an attempt to slow down the clients.  Again I'm not married to this idea and I'm open to suggestions.  I've been playing with bigqueue, maybe it can be used to pick up the slack during heavy load.

Also if you are going to do any pull requests branch off from develop not master.

**Comment by salex89** [comment]
Ok, now I understand better why it isn't really simple.

I haven't seen that Kairos is actually trying to restore the data. When the exception appears I never see the complete data in the database, and I waited. Is that a bug, then? 

KairosDB is actually doing something that is my practice, only on a smaller scale. I put a queue/broker between my web service and a worker storing the data. The service pushes data to the queue while the worker does some basic processing and stores it into Cassandra. This means I can have some latency between the data entering the queue and finishing in Cassandra which can get complicated to monitor. I want to make the process more idiomatic with KairosDB  (which means less logic in the worker). 

It would be nice if Kairos can act as a queue I am currently using. Maybe the bigqueue can to that, but again you have to provide feedback and latency information. A naive client might think everything is going great and fast while the data gets stored after 20 minutes. It boils down to the same situation I am having.

Another solution I might propose is by clearly making the calls asynchronous and provide a way to check the progress of that concrete job or a batch, using a mechanism like Java Futures. Currently you are putting multiple jobs into the same buffer, so that may be an issue. Does that look legitimate?

**Comment by gmucha** [comment]
Getting back to this - I think there should at least be some metric on lost writes - true, there are some Hector JMX metrics on write failures - but other than this, there's no way to check if C* is up to date with writes. Maybe this can be tied to BigQueue implementation - this way failing writes / queue filling up could result in backpressure - with the REST API returning something like HTTP 503 (no sense accepting writes if they're going to be lost). My preference is to let clients know that we can't keep up so that they can implement some fallback policy - retry, forget or save to some other storage.

**Comment by brianhks** [comment]
With the conversion to CQL I'm adding a queue in Kairos.  Kairos will respond only after the data is in the file backed queue for durability.

**Comment by LeiRui** [comment]
Is there any update on this issue?
